File: Android/scenes/voice/voice/src/main/java/io/agora/scene/voice/ui/fragment/ChatroomHandsDialog.java
package io.agora.scene.voice.ui.fragment;

import android.graphics.Typeface;
import android.os.Bundle;
import android.util.TypedValue;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.viewpager2.adapter.FragmentStateAdapter;
import androidx.viewpager2.widget.ViewPager2;

import com.google.android.material.imageview.ShapeableImageView;
import com.google.android.material.tabs.TabLayout;
import com.google.android.material.tabs.TabLayoutMediator;
import com.google.android.material.textview.MaterialTextView;

import java.util.ArrayList;
import java.util.Map;

import io.agora.scene.voice.R;
import io.agora.voice.baseui.BaseInitFragment;
import io.agora.voice.baseui.dialog.BaseSheetDialog;
import io.agora.voice.buddy.tool.LogTools;
import io.agora.voice.buddy.tool.DeviceTools;
import io.agora.scene.voice.databinding.VoiceRoomHandLayoutBinding;

public class ChatroomHandsDialog extends BaseSheetDialog<VoiceRoomHandLayoutBinding> {
    private int[] titles = {R.string.voice_room_raised_hands_title, R.string.voice_room_invite_hands_title};
    private ArrayList<BaseInitFragment> fragments = new ArrayList();
    private MaterialTextView title;
    private int index;
    private int mCount;
    private String roomId;
    private Bundle bundle = new Bundle();
    private ChatroomRaisedHandsFragment raisedHandsFragment;
    private ChatroomInviteHandsFragment inviteHandsFragment;
    private VoiceRoomHandLayoutBinding binding;

    private final String TAG = ChatroomHandsDialog.class.getSimpleName();

    @Nullable
    @Override
    protected VoiceRoomHandLayoutBinding getViewBinding(@NonNull LayoutInflater inflater, @Nullable ViewGroup container) {
        return VoiceRoomHandLayoutBinding.inflate(inflater, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        if (getBinding() != null)
            binding = getBinding();
        setOnApplyWindowInsets(binding.getRoot());
        initArgument();
        initView(savedInstanceState);
        initListener();
    }

    public static ChatroomHandsDialog getNewInstance() {
        return new ChatroomHandsDialog();
    }

    public void initArgument() {
        if (getArguments() != null && getArguments().containsKey("roomId"))
            roomId = getArguments().getString("roomId");
    }

    public void initView(Bundle savedInstanceState) {
        fragments.add(new ChatroomRaisedHandsFragment());
        fragments.add(new ChatroomInviteHandsFragment());
        setupWithViewPager();
    }

    public void initListener() {
        binding.tabLayout.addOnTabSelectedListener(new TabLayout.OnTabSelectedListener() {
            @Override
            public void onTabSelected(TabLayout.Tab tab) {
                if (tab.getCustomView() != null) {
                    LogTools.logD("onTabSelected：" + mCount, TAG);
                    index = tab.getPosition();
                    title = tab.getCustomView().findViewById(R.id.mtTabText);
                    ShapeableImageView tag_line = tab.getCustomView().findViewById(R.id.tab_bg);
                    ViewGroup.LayoutParams layoutParams = title.getLayoutParams();
                    layoutParams.height = (int) DeviceTools.dp2px(getActivity(), 26);
                    title.setTextSize(TypedValue.COMPLEX_UNIT_SP, 16);
                    title.setGravity(Gravity.CENTER);
                    title.setTypeface(null, Typeface.BOLD);
                    String content = getString(titles[index]) + getString(R.string.voice_room_tab_layout_count, String.valueOf(mCount));
                    title.setText(content);
                    title.setTextColor(getResources().getColor(R.color.voice_dark_grey_color_040925));
                    tag_line.setBackgroundColor(getResources().getColor(R.color.voice_color_156ef3));
                }
            }

            @Override
            public void onTabUnselected(TabLayout.Tab tab) {
                if (tab.getCustomView() != null) {
                    LogTools.logD("onTabUnselected：" + mCount, TAG);
                    MaterialTextView title = tab.getCustomView().findViewById(R.id.mtTabText);
                    ShapeableImageView tag_line = tab.getCustomView().findViewById(R.id.tab_bg);
                    title.setTextSize(TypedValue.COMPLEX_UNIT_SP, 14);
                    title.setText(titles[tab.getPosition()]);
                    title.setTypeface(null, Typeface.NORMAL);
                    title.setTextColor(getResources().getColor(R.color.voice_color_979cbb));
                    tag_line.setBackgroundColor(getResources().getColor(R.color.voice_white));
                }
            }

            @Override
            public void onTabReselected(TabLayout.Tab tab) {
                LogTools.logD("onTabReselected：", TAG);
                title = tab.getCustomView().findViewById(R.id.mtTabText);
                title.setText(titles[tab.getPosition()]);
                title.setTextColor(getResources().getColor(R.color.voice_dark_grey_color_040925));
                title.setTypeface(null, Typeface.BOLD);
            }
        });
        binding.vpFragment.setCurrentItem(0);
        binding.tabLayout.selectTab(binding.tabLayout.getTabAt(0));
    }

    private void setupWithViewPager() {
        binding.vpFragment.setOffscreenPageLimit(ViewPager2.OFFSCREEN_PAGE_LIMIT_DEFAULT);
        // set adapter
        binding.vpFragment.setAdapter(new FragmentStateAdapter(getActivity().getSupportFragmentManager(), getLifecycle()) {
            @NonNull
            @Override
            public Fragment createFragment(int position) {
                if (fragments.get(position) instanceof ChatroomRaisedHandsFragment) {
                    raisedHandsFragment = ((ChatroomRaisedHandsFragment) fragments.get(position));
                    raisedHandsFragment.setItemCountChangeListener(new ChatroomRaisedHandsFragment.itemCountListener() {
                        @Override
                        public void getItemCount(int count) {
                            mCount = count;
                            if (getActivity() != null) {
                                String content = requireActivity().getString(titles[index]) + getString(R.string.voice_room_tab_layout_count, String.valueOf(mCount));
                                LogTools.logD("getItemCount content1: " + content, TAG);
                                title.setText(content);
                            }
                        }
                    });
                } else if (fragments.get(position) instanceof ChatroomInviteHandsFragment) {
                    inviteHandsFragment = ((ChatroomInviteHandsFragment) fragments.get(position));
                    inviteHandsFragment.setItemCountChangeListener(new ChatroomInviteHandsFragment.itemCountListener() {
                        @Override
                        public void getItemCount(int count) {
                            mCount = count;
                            if (getActivity() != null) {
                                String content = requireActivity().getResources().getString(titles[index]) + getString(R.string.voice_room_tab_layout_count, String.valueOf(mCount));
                                LogTools.logD("getItemCount content2: " + content, TAG);
                                title.setText(content);
                            }
                        }
                    });
                }
                bundle.putString("roomId", roomId);
                fragments.get(position).setArguments(bundle);
                return fragments.get(position);
            }

            @Override
            public int getItemCount() {
                return titles.length;
            }

        });

        // set TabLayoutMediator
        TabLayoutMediator mediator = new TabLayoutMediator(binding.tabLayout, binding.vpFragment, new TabLayoutMediator.TabConfigurationStrategy() {
            @Override
            public void onConfigureTab(@NonNull TabLayout.Tab tab, int position) {
                tab.setCustomView(R.layout.voice_room_hands_tab_item);
                title = tab.getCustomView().findViewById(R.id.mtTabText);
                title.setText(titles[position]);
            }
        });
        // setup with viewpager2
        mediator.attach();
    }

    public void update(int index) {
        switch (index) {
            case 0:
                if (raisedHandsFragment != null)
                    raisedHandsFragment.reset();
                break;
            case 1:
                if (inviteHandsFragment != null)
                    inviteHandsFragment.reset();
                break;
        }
    }

    public void check(Map<String, String> map) {
        if (inviteHandsFragment != null)
            inviteHandsFragment.MicChanged(map);
    }

}

File: Android/scenes/voice/voice/src/main/java/io/agora/scene/voice/ui/fragment/ChatroomHandsDialog.kt
package io.agora.scene.voice.ui.fragment

import io.agora.voice.buddy.tool.LogTools.logD
import io.agora.voice.baseui.dialog.BaseSheetDialog
import io.agora.scene.voice.R
import com.google.android.material.textview.MaterialTextView
import android.os.Bundle
import android.view.LayoutInflater
import android.view.ViewGroup
import com.google.android.material.tabs.TabLayout.OnTabSelectedListener
import com.google.android.material.tabs.TabLayout
import com.google.android.material.imageview.ShapeableImageView
import android.util.TypedValue
import android.graphics.Typeface
import android.view.Gravity
import android.view.View
import androidx.fragment.app.Fragment
import androidx.viewpager2.widget.ViewPager2
import androidx.viewpager2.adapter.FragmentStateAdapter
import com.google.android.material.tabs.TabLayoutMediator
import io.agora.scene.voice.databinding.VoiceRoomHandLayoutBinding
import io.agora.voice.buddy.tool.DeviceTools
import io.agora.voice.buddy.tool.ResourcesTools

class ChatroomHandsDialog : BaseSheetDialog<VoiceRoomHandLayoutBinding>() {
    private val titles = intArrayOf(R.string.voice_room_raised_hands_title, R.string.voice_room_invite_hands_title)
    private val fragments: MutableList<Fragment> = mutableListOf()
    private var title: MaterialTextView? = null
    private var index = 0
    private var mCount = 0
    private var roomId: String? = null
    private val bundle = Bundle()
    private var raisedHandsFragment: ChatroomRaisedHandsFragment? = null
    private var inviteHandsFragment: ChatroomInviteHandsFragment? = null
    private val TAG = ChatroomHandsDialog::class.java.simpleName

    override fun getViewBinding(inflater: LayoutInflater, container: ViewGroup?): VoiceRoomHandLayoutBinding {
        return VoiceRoomHandLayoutBinding.inflate(inflater, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        roomId = arguments?.getString("roomId")
        binding?.root?.let {
            setOnApplyWindowInsets(it)
        }
        initView()
        initListener()
    }

    fun initView() {
        fragments.add(ChatroomRaisedHandsFragment())
        fragments.add(ChatroomInviteHandsFragment())
        setupWithViewPager()
    }

    fun initListener() {
        binding?.tabLayout?.addOnTabSelectedListener(object : OnTabSelectedListener {
            override fun onTabSelected(tab: TabLayout.Tab) {
                tab.customView?.let {
                    "onTabSelected：$mCount".logD(TAG)
                    index = tab.position
                    title = it.findViewById(R.id.mtTabText)
                    val tag_line = it.findViewById<ShapeableImageView>(R.id.tab_bg)
                    val layoutParams = title?.layoutParams.apply {

                    }
                    layoutParams?.height = DeviceTools.dp2px(requireContext(), 26f)
                    title?.apply {
                        setTextSize(TypedValue.COMPLEX_UNIT_SP, 16f)
                        gravity = Gravity.CENTER
                        setTypeface(null, Typeface.BOLD)
                        val content =
                            getString(titles[index]) + getString(
                                R.string.voice_room_tab_layout_count,
                                mCount.toString()
                            )
                        text = content
                        setTextColor(ResourcesTools.getColor(resources, R.color.voice_dark_grey_color_040925))
                    }

                    tag_line.setBackgroundColor(ResourcesTools.getColor(resources, R.color.voice_color_156ef3))
                }
            }

            override fun onTabUnselected(tab: TabLayout.Tab) {
                if (tab.customView != null) {
                    "onTabUnselected：$mCount".logD(TAG)
                    val title = tab.customView?.findViewById<MaterialTextView>(R.id.mtTabText)
                    val tag_line = tab.customView?.findViewById<ShapeableImageView>(R.id.tab_bg)
                    title?.apply {
                        setTextSize(TypedValue.COMPLEX_UNIT_SP, 14f)
                        setText(titles[tab.position])
                        setTypeface(null, Typeface.NORMAL)
                        setTextColor(ResourcesTools.getColor(resources, R.color.voice_color_979cbb))
                    }
                    tag_line?.setBackgroundColor(ResourcesTools.getColor(resources, R.color.voice_white))
                }
            }

            override fun onTabReselected(tab: TabLayout.Tab) {
                "onTabReselected：".logD(TAG)
                title = tab.customView?.findViewById(R.id.mtTabText)
                title?.apply {
                    setText(titles[tab.position])
                    setTextColor(ResourcesTools.getColor(resources, R.color.voice_dark_grey_color_040925))
                    setTypeface(null, Typeface.BOLD)
                }
            }
        })
        binding?.vpFragment?.currentItem = 0
        binding?.tabLayout?.selectTab(binding?.tabLayout?.getTabAt(0))
    }

    private fun setupWithViewPager() {
        binding?.vpFragment?.offscreenPageLimit = ViewPager2.OFFSCREEN_PAGE_LIMIT_DEFAULT
        // set adapter
        binding?.vpFragment?.adapter = object : FragmentStateAdapter(childFragmentManager, lifecycle) {
            override fun createFragment(position: Int): Fragment {
                if (fragments[position] is ChatroomRaisedHandsFragment) {
                    raisedHandsFragment = fragments[position] as ChatroomRaisedHandsFragment?
                    raisedHandsFragment?.setItemCountChangeListener(object :ChatroomRaisedHandsFragment.ItemCountListener{
                        override fun getItemCount(count: Int) {
                            mCount = count
                            activity?.let {
                                val content = requireActivity().getString(titles[index]) + getString(
                                    R.string.voice_room_tab_layout_count,
                                    mCount.toString()
                                )
                                "getItemCount content1: $content".logD(TAG)
                                title?.text = content
                            }
                        }

                    })
                } else if (fragments[position] is ChatroomInviteHandsFragment) {
                    inviteHandsFragment = fragments[position] as ChatroomInviteHandsFragment?
                    inviteHandsFragment?.setItemCountChangeListener(object :
                        ChatroomInviteHandsFragment.itemCountListener {
                        override fun getItemCount(count: Int) {
                            mCount = count
                            if (activity != null) {
                                val content = requireActivity().resources.getString(titles[index]) + getString(
                                    R.string.voice_room_tab_layout_count,
                                    mCount.toString()
                                )
                                "getItemCount content2: $content".logD(TAG)
                                title?.text = content
                            }
                        }
                    })
                }
                bundle.putString("roomId", roomId)
                fragments[position].arguments = bundle
                return fragments[position]
            }

            override fun getItemCount(): Int {
                return titles.size
            }
        }


        binding?.apply {
            // set TabLayoutMediator
            val mediator = TabLayoutMediator(tabLayout, vpFragment) { tab, position ->
                tab.setCustomView(R.layout.voice_room_hands_tab_item)
                title = tab.customView?.findViewById(R.id.mtTabText)
                title?.setText(titles[position])
            }
            // setup with viewpager2
            mediator.attach()
        }

    }

    fun update(index: Int) {
        when (index) {
            0 -> raisedHandsFragment?.reset()
            1 -> inviteHandsFragment?.reset()
        }
    }

    fun check(map: Map<String, String>) {
        inviteHandsFragment?.micChanged(map)
    }

    companion object {
        @JvmStatic
        val newInstance: ChatroomHandsDialog
            get() = ChatroomHandsDialog()
    }
}